frame/assistants/shadow-liketh/src/shadow.rs

```diff
@@ -0,0 +1,189 @@
+            .get_file(format!("{}/{}", dir, file))
+            .or_else(|| GRAPHQL_DIR.get_file(format!("generic/{}", file)))
+            .ok_or_else(|| ShadowComponentError::GraphQL("No graphql fround".to_string()))?;
+        graph
+            .contents_utf8()
+            .ok_or_else(|| ShadowComponentError::GraphQL("Failed to read graphql file".to_string()))
+    }
+}
+
+impl Shadow {
+    pub async fn parcel(&self, block_number: u64) -> ShadowComponentReuslt<HeaderParcel> {
+        let header = self
+            .eth
+            .get_header_by_number(block_number)
+            .await
+            .map_err(|e| ShadowComponentError::Ethereum(format!("{:?}", e)))?;
+        let mmr_root = self.mmr_root(block_number).await?;
+        Ok(HeaderParcel { mmr_root, header })
+    }
+
+    pub async fn receipt(
+        &self,
+        tx: impl AsRef<str>,
+        last: u64,
+    ) -> ShadowComponentReuslt<EthereumReceiptWithMMRProof> {
+        let resp = self
+            .http
+            .get(&format!(
+                "{}/ethereum/receipt/{}/{}",
+                &self.config.endpoint,
+                tx.as_ref(),
+                last
+            ))
+            .send()
+            .await?;
+        if resp.status() == reqwest::StatusCode::INTERNAL_SERVER_ERROR {
+            return Err(ShadowComponentError::InternalServer(resp.text().await?));
+        }
+        let result: serde_json::Value = resp.json().await?;
+        if let Some(err) = result.get("error") {
+            let msg = err.as_str().ok_or_else(|| {
+                ShadowComponentError::Cusom("Failed parse error message".to_string())
+            })?;
+            return Err(ShadowComponentError::Cusom(msg.to_owned()));
+        }
+        let receipt: EthereumReceiptJson = serde_json::from_value(result)?;
+        let header = &receipt.header;
+
+        let (member_leaf_index, last_leaf_index) = (header.number, last - 1);
+        let proof = self.mmr_proof(member_leaf_index, last_leaf_index).await?;
+        let mmr_proof = MMRProofJson {
+            member_leaf_index,
+            last_leaf_index,
+            proof,
+        };
+        Ok(EthereumReceiptWithMMRProof { receipt, mmr_proof })
+    }
+}
+
+impl Shadow {
+    fn calc_peaks(
+        &self,
+        positions: Vec<MMRPosition>,
+    ) -> ShadowComponentReuslt<Vec<(u64, [u8; 32])>> {
+        positions
+            .iter()
+            .map(|item| {
+                let array: arrayvec::ArrayVec<_, 32> = item.hash.clone().into_iter().collect();
+                match array
+                    .into_inner()
+                    .map_err(|e| ShadowComponentError::MMR(format!("{:?}", e)))
+                {
+                    Ok(v) => Ok((item.id, v)),
+                    Err(e) => Err(e),
+                }
+            })
+            .collect::<Vec<ShadowComponentReuslt<(u64, [u8; 32])>>>()
+            .into_iter()
+            .collect::<ShadowComponentReuslt<Vec<(u64, [u8; 32])>>>()
+    }
```

Comment:
There are more clean ways to implement this? @AurevoirXavier

Reply:
https://docs.rs/serde-hex/latest/serde_hex/

```
#[derive(Debug,Serialize,Deserialize)]
struct Foo {
   #[serde(with = "SerHex::<StrictPfx>")]
   bar: [u8;32]
}
```
the serde-hex crate is support deserialize to a fixed array

Reply:
https://github.com/darwinia-network/bridger/pull/429/commits/e929f66c344bcecf9f19392dece7ffa8025ad522